numgrid/AtomGrid.cpp

```diff
@@ -0,0 +1,284 @@
+            num_angular = get_closest_num_angular(num_angular);
+            if (num_angular < min_num_angular_points_closest)
+                num_angular = min_num_angular_points_closest;
+        }
+
+        int angular_off = get_angular_order(num_angular) * MAX_ANGULAR_GRID;
+
+        for (int iang = 0; iang < num_angular; iang++)
+        {
+            atom_grid_x_au.push_back(angular_x[angular_off + iang] * radial_r);
+            atom_grid_y_au.push_back(angular_y[angular_off + iang] * radial_r);
+            atom_grid_z_au.push_back(angular_z[angular_off + iang] * radial_r);
+
+            atom_grid_w.push_back(4.0 * M_PI * angular_w[angular_off + iang] *
+                                  radial_w);
+
+            num_grid_points++;
+        }
+    }
+
+    delete[] angular_x;
+    delete[] angular_y;
+    delete[] angular_z;
+    delete[] angular_w;
+}
+
+void numgrid_free_atom_grid(context_t *context)
+{
+    if (!context)
+        return;
+    delete AS_TYPE(AtomGrid, context);
+}
+AtomGrid::~AtomGrid()
+{
+    atom_grid_x_au.clear();
+    atom_grid_y_au.clear();
+    atom_grid_z_au.clear();
+    atom_grid_w.clear();
+
+    radial_atom_grid_r_au.clear();
+    radial_atom_grid_w.clear();
+}
+
+int numgrid_get_num_grid_points(const context_t *context)
+{
+    return AS_CTYPE(AtomGrid, context)->get_num_grid_points();
+}
+int AtomGrid::get_num_grid_points() const { return num_grid_points; }
+
+int numgrid_get_num_radial_grid_points(const context_t *context)
+{
+    return AS_CTYPE(AtomGrid, context)->get_num_radial_grid_points();
+}
+int AtomGrid::get_num_radial_grid_points() const
+{
+    return num_radial_grid_points;
+}
+
+void numgrid_get_grid(const context_t *context,
+                      const int num_centers,
+                      const int center_index,
+                      const double x_coordinates_au[],
+                      const double y_coordinates_au[],
+                      const double z_coordinates_au[],
+                      const int proton_charges[],
+                      double grid_x_au[],
+                      double grid_y_au[],
+                      double grid_z_au[],
+                      double grid_w[])
+{
+    return AS_CTYPE(AtomGrid, context)
+        ->get_grid(num_centers,
+                   center_index,
+                   x_coordinates_au,
+                   y_coordinates_au,
+                   z_coordinates_au,
+                   proton_charges,
+                   grid_x_au,
+                   grid_y_au,
+                   grid_z_au,
+                   grid_w);
+}
+void AtomGrid::get_grid(const int num_centers,
+                        const int center_index,
+                        const double x_coordinates_au[],
+                        const double y_coordinates_au[],
+                        const double z_coordinates_au[],
+                        const int proton_charges[],
+                        double grid_x_au[],
```

Comment:
Are we sure these `grid_*` arrays do not contain any crap when entering this getter? In case you want to zero them out on entrance, you can use `std::fill_n(&grid_x_au, num_grid_points, 0.0)`.

Reply:
Good idea.